### PR TITLE
[FLINK-32314][rpc] Ignore classloading errors after actorsystem shutdown 

### PR DESCRIPTION
**Pull request by zentol**
Stray threads in akka/netty can run into classloading errors because they aren't shut down (for some reason) despite the actor system having shut down.
Ignore errors in this specific situation.

This is primarily relevant for testing.

### CODE REVIEW

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java [hunk 1]
```diff
@@ -85,17 +86,52 @@ private static RobustActorSystem create(
...
+        }
+
+        public void notifyShutdownComplete() {
+            shutdownComplete.set(true);
+        }
+
+        @Override
+        public void uncaughtException(Thread t, Throwable e) {
+            if (shutdownComplete.get()
+                    && (e instanceof NoClassDefFoundError || e instanceof ClassNotFoundException)) {
```
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java:124]: I'm wondering whether we should make this more restrictive (checking that the thread's name includes the `akka` keyword). So far, we have seen only `flink-akka` and `flink-metrics-akka` as the current thread in issues reported under FLINK-32189.

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java:1]: It's not part of the change but why do we have to declare `RobustActorSystem` as abstract and overwriting the `uncaughtExceptionHandler()` instead of setting the exception handler through the constructor and implementing the method properly? :thinking:

File: flink-rpc/flink-rpc-akka/src/test/java/org/apache/flink/runtime/rpc/akka/RobustActorSystemTest.java [hunk 2]
```diff
@@ -84,6 +84,34 @@ void testUncaughtExceptionHandlerFromActor() {
...
+    void testHonorClassloadingErrorBeforeShutdown() {
+        robustActorSystem
+                .uncaughtExceptionHandler()
+                .uncaughtException(Thread.currentThread(), new NoClassDefFoundError(""));
+
+        assertThat(testingUncaughtExceptionHandler.findUncaughtExceptionNow()).isPresent();
+    }
+
+    @Test
+    void testIgnoreClassloadingErrorAfterShutdown() {
```
Comment (XComp) [inline flink-rpc/flink-rpc-akka/src/test/java/org/apache/flink/runtime/rpc/akka/RobustActorSystemTest.java:97]: nit: this test could be split up in two test methods (or through a parameterized test), e.g.:

```
    @ParameterizedTest
    @ValueSource(classes = {NoClassDefFoundError.class, ClassNotFoundException.class})
    void testIgnoreClassloadingErrorAfterShutdown(Class<? extends Throwable> clazz)
            throws InstantiationException, IllegalAccessException {
        // wait for termination
        robustActorSystem.terminate();
        robustActorSystem.getWhenTerminated().toCompletableFuture().join();

        robustActorSystem
                .uncaughtExceptionHandler()
                .uncaughtException(Thread.currentThread(), clazz.newInstance());

        assertThat(testingUncaughtExceptionHandler.findUncaughtExceptionNow()).isEmpty();
    }
```

**Review by XComp — APPROVED** [review]
I still feel like we should close FLINK-32314 as a duplicate of FLINK-32189, relabel this PR accordingly and document the reasoning for this change in FLINK-32189 (why we don't proceed further with finding the cause) since we collected most of the related test instabilities there.

I feel like we should be quite careful with documenting this specific issue properly.

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java (hunk 1 above)
Comment (zentol) [reply flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java:124]: I can do that, but this uncaught exception handler only applies to the akka actor system, so it should implicitly only cover akka threads (independent of how they are named).

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java
Comment (XComp) [reply flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java:1]: That's what I struggled with. But I might just struggle to find the method call in the scala code :thinking: It's probably the `threadFactory` instantiation that the JavaDoc is referring to.

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java (hunk 1 above)
Comment (XComp) [reply flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java:124]: ah ok, I actually wanted to check where the class is used but forgot. You're right - it doesn't make a difference. then.

File: flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java
Comment (zentol) [reply flink-rpc/flink-rpc-akka/src/main/java/org/apache/flink/runtime/rpc/akka/RobustActorSystem.java:1]: > It's probably the threadFactory instantiation that the JavaDoc is referring to.

yes
